core/src/main/java/net/adoptopenjdk/icedteaweb/resources/downloader/PrioritizedParallelExecutor.java

```diff
@@ -0,0 +1,220 @@
+ * with higher priority (if any) have completed exceptionally
+ */
+public class PrioritizedParallelExecutor {
+
+    private final ExecutorService executor;
+
+    /**
+     * Constructor
+     *
+     * @param executor the executor service which is used for executing the callables.
+     */
+    public PrioritizedParallelExecutor(ExecutorService executor) {
+        this.executor = executor;
+    }
+
+    /**
+     * Executes the passed callables in parallel
+     * The callables must be sorted by priority with the highest priority as the first in list.
+     * Executions of lower priority are canceled as soon as a higher priority result has successfully been calculated.
+     * The returned {@link Future} will complete as soon as the first callable completed successfully and all other
+     * callables with higher priority (if any) have completed exceptionally.
+     *
+     * @param callables the callables to execute.
+     * @return a future holding the best result from all callables.
+     */
+    public <V> Future<V> getSuccessfulResultWithHighestPriority(List<Callable<V>> callables) {
+        if (callables == null || callables.isEmpty()) {
+            throw new IllegalArgumentException("No callables");
+        }
+
+        if (callables.size() == 1) {
+            return executor.submit(callables.get(0));
+        }
+
+        final CancelingCallable<V> head = new CancelingCallable<>(callables.get(0));
+        final List<Callable<V>> tail = callables.subList(1, callables.size());
+
+        final ChainableFuture<V> result = new ChainableFuture<>(executor.submit(head));
+        final Future<V> lowerPriority = getSuccessfulResultWithHighestPriority(tail);
+
+        head.setLowerPriority(lowerPriority);
+        result.setLowerPriority(lowerPriority);
+
+        return result;
+    }
+
+    /**
+     * Wrapper around {@link Callable} which will cancel any lower priority after completing the call successfully.
+     */
+    private static class CancelingCallable<V> implements Callable<V> {
+
+        private final AtomicBoolean complete = new AtomicBoolean(false);
+        private final AtomicReference<Future<V>> lowerPriority = new AtomicReference<>();
+
+        private final Callable<V> delegate;
+
+        private CancelingCallable(Callable<V> delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Set a lower priority to be canceled.
+         * This is not passed in the constructor as the value is not known then.
+         * <p>
+         * If this callable has already been completed successfully then the passed future will be canceled immediately.
+         *
+         * @param lowerPriority the lower priority future to be canceled after completing this callable successfully.
+         */
+        void setLowerPriority(Future<V> lowerPriority) {
+            this.lowerPriority.set(lowerPriority);
+            if (complete.get()) {
+                cancelLowerPriority();
```

Comment:
Since cancelLowerPriority calls cancel() on the future object the method might block (cancel() blocks until the future is really canceled)

Reply:
yes, so should I execute it in a different thread then??

Reply:
the best case would be to not start it if this one is already completed. And this check must be locked.

But this is again the problem. As far as I can say this complete PR can end in several problems based on race conditions that we just do not see at the moment.